ls1intum/Artemis
# `Programming exercises`: Fix an issue with excess submissions with Gitlab

### Checklist
#### General


#### Server

### Motivation and Context


Closes #5145 
@dfuchss reported in #5145 that sometimes Artemis produces multiple submissions with different commits.
After investigation, I found out that this was the case if the student submits multiple commits within one push which happens rather rarely.

### Description

Gitlab provides in their [documentation ](https://docs.gitlab.com/ee/user/project/integrations/webhook_events.html#push-events) an example of the payload of the push event webhook. The commits are sorted here in ascending order.

In our tests, however, we assumed a descending order. Because the previous implementation only took the first element on push-trigger, it always took the first commit, not the latest.

Gitlab provides the latest commit hash in a dedicated field. I took this hash and search for this specific commit. If it exists, which it usually does, we fill the response container with the details, otherwise, it only contains the hash itself.

@dfuchss confirmed that this fix resolves the issues he had.

Additionally, I added some test cases for this method and renamed the test request files in a more intuitive manner.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Programming Exercise
- Gitlab/Jenkins Setup

1. Create locally two commits and push both at the same time.
2. Make sure that only one submission got created with the commit hash of the latest commit

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| GitlabService.java | 50% | 83.66% |
